### PR TITLE
feat: add markdown-it plugin to collect headings

### LINKED ISSUE #51: Inject markdown headers as page attributes
[Issuehunt badges](https://issuehunt.io/r/egoist/saber/issues/51)

Add a markdown-it plugin to inject markdown headers `[{ text: 'Hello', slug: hello' }]` as `page.attributes.markdownHeaders`, similar to Vuepress's `this.$page.headers`.

When to inject:

- `page.attributes.injectMarkdownHeaders === true`
- Or `page.attributes.injectMarkdownHeaders !== false && api.config.markdown.injectMarkdownHeaders`

The injected markdown headers can be used by https://github.com/egoist/saber/pull/49

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/egoist/saber/issues/51).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/egoist/saber/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by krmax44**
Injects all Markdown headings to `page.attributes`. I used [this plugin](https://github.com/medfreeman/markdown-it-toc-and-anchor) for some base code. Closes #51.

Headings will be injected when:
  - `page.attributes.injectMarkdownHeaders === true`
  - Or `page.attributes.injectMarkdownHeaders !== false && api.config.markdown.injectMarkdownHeaders`

`page.attributes.markdownHeadings` is an array containg objects like `{ text: 'Foo', slug: 'foo' }`.

Usage:

```yml
# saber-config.yml

markdown:
  options:
    injectMarkdownHeadings: true
```

or in individual Markdown files:
```md
---
title: Foo
injectMarkdownHeadings: true
---

# cool
## stuff
### going
#### on
```

### CODE REVIEW

**Comment by MarchWorks** [comment]
markdown-it plugin should be here I think https://github.com/saberland/saber/tree/master/packages/saber/lib/markdown

**Comment by krmax44** [comment]
> markdown-it plugin should be here I think https://github.com/saberland/saber/tree/master/packages/saber/lib/markdown
> with there test in some cases

yup makes more sense in there! changed it

**Comment by egoist** [comment]
maybe add a test case?

**Comment by krmax44** [comment]
I also removed all the semi-good regex based slug functions and replaced it with slugo. Just 255B minified and gzipped :drooling_face:

**Comment by krmax44** [comment]
@egoist is anything missing here or can it be merged?

**Comment by egoist** [comment]
I think we should support selecting headings with specific levels as well: `{injectMarkdownHeadings: {slectedLevels: [2,3,4]}}

Regarding this, maybe we can implement the headings plugin in the anchor plugin directly? Since many features are overlapping.

File: packages/saber/lib/markdown/headings-plugin.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            text = heading.children[1].content
+          } else {
+            text = heading.content
+          }
+
+          const slug = slugFn(text)
+
+          headings.push({
+            text,
+            slug
```
Comment (egoist) [inline packages/saber/lib/markdown/headings-plugin.js]: let's inject heading level as well.

**Comment by krmax44** [comment]
@egoist so you mean not only add the heading's level to the output but also let the user choose which one to inject at all?

**Comment by egoist** [comment]
yes, by default we inject h2~h5

**Comment by krmax44** [comment]
@egoist anything else to add?

**Comment by egoist** [comment]
> Regarding this, maybe we can implement the headings plugin in the anchor plugin directly? Since many features are overlapping.

thoughts on this?

**Comment by krmax44** [comment]
> > Regarding this, maybe we can implement the headings plugin in the anchor plugin directly? Since many features are overlapping.
> 
> thoughts on this?

Makes sense, I'll merge them together.

**Comment by krmax44** [comment]
@egoist I'm unsure if the possibility of in/excluding certain levels for heading injection makes that much sense. I'd guess it just throws off most people expecting that `injectMarkdownHeadings = true` means inject everything. It's also not really costing anything to just inject all levels and I hardly see a use-case for it, so the option just feels sort of unnecessary to me. If you really need that functionality, you can easily add it:  `page.attributes.injectMarkdownHeadings.filter(({ level } => [2, 4, 5].includes(level))`

**Comment by egoist** [comment]
I guess you're right 😅

File: packages/saber/lib/plugins/transformer-markdown.js [hunk 2]
```diff
@@ -80,10 +80,10 @@ function transformMarkdown(api, page) {
       name: 'headings',
       resolve: require.resolve('../markdown/headings-plugin'),
       options: {
+        ...markdown.headings,
```
Comment (krmax44) [inline packages/saber/lib/plugins/transformer-markdown.js]: Bit dangerous, what if `headings` is undefined? Also validateConfig will throw an error on start if you use this option in `saber-config`.

Comment (egoist) [reply packages/saber/lib/plugins/transformer-markdown.js]: `undefined` is fine.

Comment (krmax44) [reply packages/saber/lib/plugins/transformer-markdown.js]: always thought it would throw an error :eyes:

**Pull request merged**